shell/platform/darwin/ios/framework/Source/FlutterEngine.mm

```diff
@@ -498,23 +522,50 @@ - (BOOL)createShell:(NSString*)entrypoint libraryURI:(NSString*)libraryURI {
     FML_LOG(ERROR) << "Could not start a shell FlutterEngine with entrypoint: "
                    << entrypoint.UTF8String;
   } else {
-    [self setupChannels];
-    [self onLocaleUpdated:nil];
-    if (!_platformViewsController) {
-      _platformViewsController.reset(new flutter::FlutterPlatformViewsController());
-    }
-    _publisher.reset([[FlutterObservatoryPublisher alloc] init]);
-    [self maybeSetupPlatformViewChannels];
-    _shell->GetIsGpuDisabledSyncSwitch()->SetSwitch(_isGpuDisabled ? true : false);
+    [self setUpWhenInit];
   }
-
   return _shell != nullptr;
 }
 
+- (BOOL)createShell:(NSString*)entrypoint libraryURI:(NSString*)libraryURI {
+  return [self createShell:entrypoint libraryURI:libraryURI asyncInitCallback:nil];
+}
+
+- (void)setUpWhenInit {
+  [self setupChannels];
+  [self onLocaleUpdated:nil];
+  if (!_platformViewsController) {
+    _platformViewsController.reset(new flutter::FlutterPlatformViewsController());
+  }
+  _publisher.reset([[FlutterObservatoryPublisher alloc] init]);
+  [self maybeSetupPlatformViewChannels];
+  _shell->GetIsGpuDisabledSyncSwitch()->SetSwitch(_isGpuDisabled ? true : false);
+}
+
 - (BOOL)run {
   return [self runWithEntrypoint:FlutterDefaultDartEntrypoint libraryURI:nil];
 }
 
+- (void)asyncRun:(InitCallBackBlock)asyncCallback {
+  if (asyncCallback == nil) {
+    FML_LOG(WARNING) << "asyncRun: callback is nil, you should care lifecycle, and called other "
+                        "api after initCallback called";
+    asyncCallback = ^(bool success) {
+    };
+  }
+  asyncCallback = [asyncCallback copy];
+  [self createShell:FlutterDefaultDartEntrypoint
+             libraryURI:nil
+      asyncInitCallback:^(bool success) {
+        if (success) {
+          [self setUpWhenInit];
+          [self launchEngine:FlutterDefaultDartEntrypoint libraryURI:nil];
+        }
+        asyncCallback(success);
+        [asyncCallback release];
+      }];
+}
+
```

Comment:
For completeness sake, if you add async to the default entrypoint, you should add it for custom entrypoints too